### PR TITLE
IGNITE-5452

### PR DESCRIPTION
**Pull request by BiryukovVA**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * Test correctness of rollback a transaction with timeout during the grid stop.
+ */
+public class IgniteTxRollbackOnStopTest extends GridCacheAbstractSelfTest {
+    /** */
+    private static final long TEST_TIMEOUT = 60_000;
+
+    /** */
+    private static final int NODES_CNT = 3;
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: NODES

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+    /** */
+    private static final int PUT_CNT = 10_000;
+
+    /** {@inheritDoc} */
+    @Override protected int gridCount() {
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java:47]: remove method

File: modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java [hunk 3]
```diff
@@ -131,6 +130,24 @@ public void removeTimeoutObject(GridTimeoutObject timeoutObj) {
     }
 
     /**
+     * Wraps {@link TimeoutWorker} in a {@link IgniteThread}
+     * with access to {@link TimeoutWorker#cancel()} method.
+     */
+    private static class GridWorkerWrapper extends IgniteThread{
```
Comment (dream-x) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java]: private static class GridWorkerWrapper extends IgniteThread {

File: modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java [hunk 4]
```diff
@@ -273,4 +290,4 @@ public void removeTimeoutObject(GridTimeoutObject timeoutObj) {
             return S.toString(CancelableTask.class, this);
         }
     }
-}
\ No newline at end of file
+}
```
Comment (dream-x) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java:293]: new line after `}`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        return TEST_TIMEOUT;
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testTxRollbackOnStop() throws Exception {
+        IgniteEx ignite1 = grid(0);
+        IgniteEx ignite2 = grid(1);
+        IgniteEx ignite3 = startGrid(gridCount()); //to be closed.
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]:
```
for (int i = 0; i < NODES; i++)
     startGrid(i);
```
or
```
startGridsMultiThreaded(NODES);
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            U.sleep(1000);
+
+            fail("A timeout should have happened.");
+        }
+        catch (Exception e) {
+            if (!X.hasCause(e, TransactionTimeoutException.class))
+                throw e;
+        }
+
+        ignite3.close();
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]:
```
stopGrid(2);

awaitPartitionMapExchange();

info("Ignite3 closed.");
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        assertDoesNotContainLockTimeoutObjects();
+
+        logTimeoutObjectsFrequency();
+
+        //Check that the values have not changed and lock can be acquired.
+        try (Transaction tx2 = ignite2.transactions().txStart(PESSIMISTIC, SERIALIZABLE)) {
+            info("Tx2 started");
+
+            for (int i = 0; i < PUT_CNT; i++) {
+                assertEquals(cache2.get(i).intValue(), Integer.MAX_VALUE);
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: assertEquals(cache2.get(i), Integer.MAX_VALUE);

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        ignite3.close();
+        info("Ignite3 closed.");
+
+        assertDoesNotContainLockTimeoutObjects();
+
+        logTimeoutObjectsFrequency();
+
+        //Check that the values have not changed and lock can be acquired.
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: `// Check that the values have not changed and lock can be acquired.`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                if (i % (PUT_CNT / 5) == 0)
+                    logTimeoutObjectsFrequency();
+            }
+
+            tx2.commit();
+        }
+
+        info("Tx2 stopped");
+
+        //Check that that changes committed.
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: // Check that that changes committed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        for (int i = 0; i < PUT_CNT; i++)
+            assertEquals(cache1.get(i).intValue(), i);
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: assertEquals(cache1.get(i), i);

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testTxRollbackOnStop() throws Exception {
+        IgniteEx ignite1 = grid(0);
+        IgniteEx ignite2 = grid(1);
+        IgniteEx ignite3 = startGrid(gridCount()); //to be closed.
+
+        IgniteCache<Integer, Integer> cache1 = ignite1.cache(DEFAULT_CACHE_NAME);
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]:
```
IgniteCache<Integer, Integer> cache1 = grid(0).cache(DEFAULT_CACHE_NAME);
IgniteCache<Integer, Integer> cache2 = grid(1).cache(DEFAULT_CACHE_NAME);
IgniteCache<Integer, Integer> cache3 = grid(2).cache(DEFAULT_CACHE_NAME);
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+
+        ignite3.close();
+        info("Ignite3 closed.");
+
+        assertDoesNotContainLockTimeoutObjects();
+
+        logTimeoutObjectsFrequency();
+
+        //Check that the values have not changed and lock can be acquired.
+        try (Transaction tx2 = ignite2.transactions().txStart(PESSIMISTIC, SERIALIZABLE)) {
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: `try (Transaction tx2 = grid(1).transactions().txStart(PESSIMISTIC, SERIALIZABLE)) {`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+        IgniteCache<Integer, Integer> cache2 = ignite2.cache(DEFAULT_CACHE_NAME);
+        IgniteCache<Integer, Integer> cache3 = ignite3.cache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < PUT_CNT; i++)
+            cache1.put(i, Integer.MAX_VALUE);
+
+        logTimeoutObjectsFrequency();
+
+        info("Tx1 started");
+        ignite3.transactions().txStart(PESSIMISTIC, SERIALIZABLE, 1000, PUT_CNT);
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: `grid(2).transactions().txStart(PESSIMISTIC, SERIALIZABLE, 1000, PUT_CNT);`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessorSelfTest.java [hunk 14]
```diff
@@ -608,4 +626,28 @@ public void testTimeoutSameEndTime() throws Exception {
...
+                    // which clears interrupted flag.
+                    Thread.sleep(10_000);
+                }
+                catch (InterruptedException ignore) {
+                    // Eat exception.
+                }
+            }
+        });
+
+        ignite.close();
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessorSelfTest.java]: stopGrid(0);

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+
+import static org.apache.ignite.transactions.TransactionConcurrency.PESSIMISTIC;
+import static org.apache.ignite.transactions.TransactionIsolation.SERIALIZABLE;
+
+/**
+ * Test correctness of rollback a transaction with timeout during the grid stop.
+ */
+public class IgniteTxRollbackOnStopTest extends GridCacheAbstractSelfTest {
+    /** */
+    private static final long TEST_TIMEOUT = 60_000;
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: remove

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+    private static final int PUT_CNT = 10_000;
+
+    /** {@inheritDoc} */
+    @Override protected int gridCount() {
+        return NODES_CNT;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected long getTestTimeout() {
+        return TEST_TIMEOUT;
```
Comment (dream-x) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: return 60_000;

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java (hunk 7 above)
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: This does not work.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java (hunk 10 above)
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: This does not work.

Comment (dream-x) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: sorry :)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java (hunk 7 above)
Comment (dream-x) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteTxRollbackOnStopTest.java]: sorry :)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java (hunk 4 above)
Comment (dream-x) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessor.java:293]: https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-A.ConfigureIntelliJIDEAcodestyle

File: modules/core/src/test/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessorSelfTest.java [hunk 17]
```diff
@@ -608,4 +626,26 @@ public void testTimeoutSameEndTime() throws Exception {
...
+    public void testCancelingWithClearedInterruptedFlag() throws Exception {
+        startGrid(0).context().timeout().addTimeoutObject(new GridTimeoutObjectAdapter(10) {
+            @Override public void onTimeout() {
+                try {
+                    // Wait for CacheProcessor has stopped and cause InterruptedException
+                    // which clears interrupted flag.
+                    Thread.sleep(10_000);
+                }
+                catch (InterruptedException ignore) {
+                    // Eat exception.
```
Comment (SharplEr) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/timeout/GridTimeoutProcessorSelfTest.java]: Лучше // No-op.

**Pull request closed without merging**